interactions/utils/abc/base_context_managers.py

```diff
@@ -0,0 +1,63 @@
+from abc import ABCMeta, abstractmethod
+from asyncio import Task, get_running_loop
+from typing import TYPE_CHECKING, Optional, TypeVar, Union
+
+_T = TypeVar("_T")
+
+__all__ = ("BaseAsyncContextManager", "BaseContextManager")
+
+if TYPE_CHECKING:
+    from ...api.http.client import HTTPClient
+    from ...api.models.misc import Snowflake
+
+
+class BaseAsyncContextManager(metaclass=ABCMeta):
+    """A base class for async iterators."""
+
+    # I don't want to make it subclass the BaseContextManager since it forces implementation of __enter__ and __exit__
+
+    @abstractmethod
+    def __init__(
+        self,
+        obj: Union[int, str, "Snowflake", _T],
+        _client: Optional["HTTPClient"] = None,
+    ):
+
+        try:
+            self.loop = get_running_loop()
+        except RuntimeError as e:
+            raise RuntimeError("No running event loop detected!") from e
+
+        self.object_id = int(obj) if not hasattr(obj, "id") else int(obj.id)
+        self._client = _client
+        self.__task: Optional[Task] = None
+
+    @abstractmethod
+    def __await__(self) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    async def do_action(self):
```

Comment:
`do_action` is a rather ambiguous name. If it's required for *every* context manager, why isn't it in the sync context manager? If it's not required, why is it in the base class?

Reply:
I guess I can make it not abstract, it's intended for background tasks (like typing)